### PR TITLE
Allow rpc.remote to create RRef on self

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29543 [WIP] Add end-to-end test for RNN Module
* #29781 Always include autograd context id in rpc/remote requests
* **#29634 Allow rpc.remote to create RRef on self**

This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

### CODE REVIEW

File: test/rpc_test.py [hunk 1]
```diff
@@ -237,6 +237,12 @@ def test_self_add(self):
         self.assertEqual(fut.wait(), torch.ones(2, 2) + 1)
         self.assertEqual(ret, torch.ones(2, 2) + 1)
 
+    @dist_init
```
Comment (aazzolini) [inline test/rpc_test.py]: We also need to test that we can pass RRef's as arguments to local sync_rpc, async_rpc and remote calls.

File: torch/csrc/distributed/rpc/python_functions.cpp [hunk 2]
```diff
@@ -171,29 +178,46 @@ PyRRef pyRemotePythonUdf(
...
+        dst,
+        std::move(*pythonRemoteCall).toMessage(),
+        true /*forceGradRecording*/
+    );
+
+    ctx.addPendingUser(userRRef->forkId(), userRRef);
+    fm->addCallback(finishAcceptUserRRef);
+    return PyRRef(userRRef);
+  } else {
+    auto ownerRRef = ctx.createOwnerRRef<py::object>(true /* retainInContext*/);
```
Comment (aazzolini) [inline torch/csrc/distributed/rpc/python_functions.cpp]: retainInContext is hard to understand. would it make sense to make this a separate function call?

Comment (mrshenli) [reply torch/csrc/distributed/rpc/python_functions.cpp]: Sure, I am working on refactoring this PR, will send out an update today.

File: test/rpc_test.py (hunk 1 above)
Comment (mrshenli) [reply test/rpc_test.py]: yes, will add more tests

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 3]
```diff
@@ -96,7 +96,11 @@ Message RequestCallbackImpl::processRpc(
 
       ownerRRef->setValue(
           PythonRpcHandler::getInstance().runPythonUDF(prc.serializedPyObj()));
-      ctx.addForkOfOwner(rrefId, forkId);
+
+      if (rrefId != forkId) {
+        // This is not owner calling to self
```
Comment (aazzolini) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Can you add a more specific comment here on where we create a fork where rrefId == forkId (namely , when calling "remote" to self).

File: torch/csrc/distributed/rpc/rref_context.cpp [hunk 4]
```diff
@@ -142,7 +142,7 @@ template std::shared_ptr<OwnerRRef<py::object>> RRefContext::
 
 template <typename T>
 std::shared_ptr<OwnerRRef<T>> RRefContext::createOwnerRRef() {
-  // Don't add this OnwerRRef to the owners_ map yet, otherwise
+  // Don't add this OnwerRRef to the owners_ map yet by default, otherwise
```
Comment (aazzolini) [inline torch/csrc/distributed/rpc/rref_context.cpp]: is this change in the comment still pertinent in the latest version?

**Review by aazzolini — APPROVED** [review]
Looks good. In the future we should probably bypass serialization and deserialization and simply short-circuit the local call, but that will require access to the thread pool.

File: torch/csrc/distributed/rpc/rref_context.cpp (hunk 4 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/rref_context.cpp]: This change should be removed, let me update. Thanks!

**Pull request closed without merging**